plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/MigrateProcedure.java

```diff
@@ -0,0 +1,350 @@
+        }
+        catch (ReflectiveOperationException e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    @Inject
+    public MigrateProcedure(
+            TrinoCatalogFactory catalogFactory,
+            HiveMetastoreFactory metastoreFactory,
+            TrinoFileSystemFactory fileSystemFactory,
+            TypeManager typeManager,
+            IcebergConfig icebergConfig,
+            IcebergSecurityConfig securityConfig)
+    {
+        this.catalogFactory = requireNonNull(catalogFactory, "catalogFactory is null");
+        this.metastoreFactory = requireNonNull(metastoreFactory, "metastoreFactory is null");
+        this.fileSystemFactory = requireNonNull(fileSystemFactory, "fileSystemFactory is null");
+        this.typeManager = requireNonNull(typeManager, "typeManager is null");
+        this.formatVersion = icebergConfig.getFormatVersion();
+        this.isUsingSystemSecurity = securityConfig.getSecuritySystem() == SYSTEM;
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                "system",
+                "migrate",
+                ImmutableList.of(
+                        new Procedure.Argument("SCHEMA", VARCHAR),
+                        new Procedure.Argument("TABLE", VARCHAR)),
+                MIGRATE.bindTo(this));
+    }
+
+    public void migrate(ConnectorSession session, String schemaName, String tableName)
+    {
+        // this line guarantees that classLoader that we stored in the field will be used inside try/catch
+        // as we captured reference to PluginClassLoader during initialization of this class
+        // we can use it now to correctly execute the procedure
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            doMigrate(session, schemaName, tableName);
+        }
+    }
+
+    public void doMigrate(ConnectorSession session, String schemaName, String tableName)
+    {
+        SchemaTableName sourceTableName = new SchemaTableName(schemaName, tableName);
+        TrinoCatalog catalog = catalogFactory.create(session.getIdentity());
+        HiveMetastore metastore = metastoreFactory.createMetastore(Optional.of(session.getIdentity()));
+
+        io.trino.plugin.hive.metastore.Table hiveTable = metastore.getTable(schemaName, tableName).orElseThrow(() -> new TableNotFoundException(sourceTableName));
+        String transactionalProperty = hiveTable.getParameters().get(TRANSACTIONAL);
+        if (parseBoolean(transactionalProperty)) {
+            throw new TrinoException(NOT_SUPPORTED, "Migrating transactional tables is unsupported");
+        }
+        if (!hiveTable.getTableType().equalsIgnoreCase(MANAGED_TABLE.name())) {
+            throw new TrinoException(NOT_SUPPORTED, "The procedure supports migrating only managed tables: " + hiveTable.getTableType());
```

Comment:
Why exclude external tables?

Reply:
There's no strong reason. Just for starting with less support.